File: dotnet/src/Pactman.NonprofitCheckPlus/Configuration/ConfigResolver.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Runtime.InteropServices;
using System.Text;
using Pactman.NonprofitCheckPlus.Exceptions;

namespace Pactman.NonprofitCheckPlus.Configuration
{
    /// <summary>
    /// Validates and resolves constructor options.
    /// </summary>
    /// <remarks>
    /// Everything unusable is rejected here, at construction, before any network call is
    /// attempted.
    /// </remarks>
    public static class ConfigResolver
    {
        /// <summary>Validates and trims an API key.</summary>
        /// <param name="apiKey">The key as supplied.</param>
        /// <returns>The trimmed key.</returns>
        /// <exception cref="PactmanConfigurationException">The key is missing or blank.</exception>
        public static string ApiKey(string? apiKey)
        {
            if (apiKey is null)
            {
                throw new PactmanConfigurationException(
                    "A Pactman API key is required. Pass ApiKey, for example from "
                    + "Environment.GetEnvironmentVariable(\"PACTMAN_API_KEY\").");
            }

            var trimmed = apiKey.Trim();

            if (trimmed.Length == 0)
            {
                throw new PactmanConfigurationException(
                    "The Pactman API key is empty. Check that the environment variable holding it is set.");
            }

            return trimmed;
        }

        /// <summary>Resolves client options into a validated configuration.</summary>
        /// <param name="options">The options as supplied.</param>
        /// <returns>The fully-resolved configuration.</returns>
        /// <exception cref="PactmanConfigurationException">
        /// An unknown environment, a malformed base URL, or a nonsensical numeric option.
        /// </exception>
        public static ClientConfig Resolve(PactmanClientOptions options)
        {
            if (options is null)
            {
                throw new ArgumentNullException(nameof(options));
            }

            // Any supplied value goes through validation, blank included. A base URL of
            // "  " means a misconfigured environment variable, and quietly falling back
            // to production is the one outcome nobody wants from that.
            var hasExplicitBaseUrl = options.BaseUrl != null;
            var environment = options.Environment ?? PactmanEnvironments.Default;

            var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

            foreach (var header in options.DefaultHeaders)
            {
                headers[header.Key] = header.Value;
            }

            return new ClientConfig(
                baseUrl: hasExplicitBaseUrl ? ValidateBaseUrl(options.BaseUrl!) : environment.BaseUrl(),
                environment: hasExplicitBaseUrl ? (PactmanEnvironment?)null : environment,
                timeout: ResolveTimeout(options.Timeout),
                retry: options.Retry ?? RetryOptions.Default,
                maxRequestsPerSecond: ResolveRequestsPerSecond(options.MaxRequestsPerSecond),
                defaultHeaders: headers,
                userAgent: BuildUserAgent());
        }

        /// <summary>
        /// Builds the <c>User-Agent</c>: <c>Pactman.NonprofitCheckPlus/&lt;version&gt; (dotnet/&lt;runtime&gt;; &lt;os&gt;)</c>.
        /// </summary>
        /// <returns>A header-safe user agent string.</returns>
        public static string BuildUserAgent() => string.Format(
            CultureInfo.InvariantCulture,
            "{0}/{1} (dotnet/{2}; {3})",
            SdkVersion.PackageName,
            SdkVersion.Version,
            HeaderSafe(RuntimeInformation.FrameworkDescription),
            HeaderSafe(RuntimeInformation.OSDescription));

        /// <summary>Validates an environment name and resolves it.</summary>
        /// <param name="name">The environment name, such as <c>production</c>.</param>
        /// <returns>The named environment.</returns>
        /// <exception cref="PactmanConfigurationException">The name is not a supported environment.</exception>
        public static PactmanEnvironment ParseEnvironment(string name)
        {
            if (PactmanEnvironments.TryParse(name, out var environment))
            {
                return environment;
            }

            throw new PactmanConfigurationException(string.Format(
                CultureInfo.InvariantCulture,
                "Unknown environment \"{0}\". Supported: {1}. Use BaseUrl to target a host that is not a named environment.",
                name,
                PactmanEnvironments.SupportedNames()));
        }

        private static string ValidateBaseUrl(string baseUrl)
        {
            var trimmed = baseUrl.Trim();

            if (trimmed.Length == 0)
            {
                throw new PactmanConfigurationException("`BaseUrl` must be a non-empty URL string.");
            }

            if (!Uri.TryCreate(trimmed, UriKind.Absolute, out var uri))
            {
                throw new PactmanConfigurationException(string.Format(
                    CultureInfo.InvariantCulture,
                    "`BaseUrl` is not a valid URL: \"{0}\". Expected something like https://entities.pactman.org.",
                    baseUrl));
            }

            if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps)
            {
                throw new PactmanConfigurationException(string.Format(
                    CultureInfo.InvariantCulture,
                    "`BaseUrl` must use http or https, received \"{0}\".",
                    uri.Scheme));
            }

            // Normalized to scheme://authority[/path] with no trailing slash, so joining a
            // path onto it can never produce a double slash.
            var authority = uri.IsDefaultPort
                ? uri.Host
                : string.Format(CultureInfo.InvariantCulture, "{0}:{1}", uri.Host, uri.Port);

            return uri.Scheme + "://" + authority + uri.AbsolutePath.TrimEnd('/');
        }

        private static TimeSpan ResolveTimeout(TimeSpan? timeout)
        {
            if (timeout is null)
            {
                return ClientConfig.DefaultTimeout;
            }

            if (timeout.Value <= TimeSpan.Zero)
            {
                throw new PactmanConfigurationException(
                    "`Timeout` must be greater than zero. There is no way to disable the timeout.");
            }

            return timeout.Value;
        }

        private static double? ResolveRequestsPerSecond(double? value)
        {
            if (value is null)
            {
                return null;
            }

            if (double.IsNaN(value.Value) || double.IsInfinity(value.Value) || value.Value <= 0)
            {
                throw new PactmanConfigurationException(
                    "`MaxRequestsPerSecond` must be a number greater than zero, or omitted.");
            }

            return value;
        }

        /// <summary>
        /// Strips whatever a runtime or OS description might contain that a header cannot.
        /// </summary>
        /// <remarks>
        /// <see cref="RuntimeInformation.OSDescription"/> is free text — on Linux it is the
        /// whole <c>uname</c> line, newlines and all. An unsanitized value here would be a
        /// header-injection seam in every outbound request.
        /// </remarks>
        private static string HeaderSafe(string value)
        {
            var safe = new StringBuilder(value.Length);

            foreach (var character in value)
            {
                if (character >= 0x20 && character < 0x7F && character != '(' && character != ')')
                {
                    safe.Append(character);
                }
                else if (character == ' ' || character == '\t' || character == '\n' || character == '\r')
                {
                    safe.Append(' ');
                }
            }

            var trimmed = safe.ToString().Trim();

            return trimmed.Length == 0 ? "unknown" : trimmed;
        }
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/config/ConfigResolver.java
package org.pactman.nonprofitcheckplus.config;

import java.net.URI;
import java.net.URISyntaxException;
import org.pactman.nonprofitcheckplus.PactmanEnvironment;
import org.pactman.nonprofitcheckplus.SdkVersion;
import org.pactman.nonprofitcheckplus.exceptions.PactmanConfigurationException;
import org.pactman.nonprofitcheckplus.http.HttpExchange;
import org.pactman.nonprofitcheckplus.http.JdkHttpExchange;

/**
 * Validates and resolves constructor options.
 *
 * <p>Every rejection happens here, at construction, rather than on the first
 * request: a client that cannot work should say so where it is built, not
 * halfway through a batch.
 */
public final class ConfigResolver {

    private ConfigResolver() {
    }

    /**
     * Validates options and fills in every default.
     *
     * @param options the caller's options.
     * @return the API key and the resolved configuration, held apart.
     * @throws PactmanConfigurationException for a missing or blank API key, an
     *         unknown environment, a malformed base URL, or a nonsensical
     *         numeric option.
     */
    public static Resolved resolve(PactmanClientOptions options) {
        if (options == null) {
            throw new PactmanConfigurationException(
                    "PactmanClient requires options containing an apiKey.");
        }

        String apiKey = validateApiKey(options.apiKey());
        String baseUrl = resolveBaseUrl(options);
        long timeoutMs = resolveTimeout(options.timeoutMs());
        Double maxRequestsPerSecond = resolveRequestsPerSecond(options.maxRequestsPerSecond());

        ClientConfig config = new ClientConfig(
                baseUrl,
                options.baseUrl() == null ? environmentOrDefault(options.environment()) : null,
                timeoutMs,
                options.retry() == null ? RetryOptions.defaults() : options.retry(),
                maxRequestsPerSecond,
                options.defaultHeaders(),
                SdkVersion.userAgent(),
                resolveExchange(options));

        return new Resolved(apiKey, config);
    }

    private static PactmanEnvironment environmentOrDefault(PactmanEnvironment environment) {
        return environment == null ? PactmanEnvironment.PRODUCTION : environment;
    }

    private static String validateApiKey(String apiKey) {
        if (apiKey == null) {
            throw new PactmanConfigurationException(
                    "A Pactman API key is required. Pass `apiKey`, for example from "
                            + "System.getenv(\"PACTMAN_API_KEY\").");
        }

        if (apiKey.trim().isEmpty()) {
            throw new PactmanConfigurationException(
                    "The Pactman API key is empty. Check that the environment variable holding "
                            + "it is set.");
        }

        return apiKey.trim();
    }

    private static String resolveBaseUrl(PactmanClientOptions options) {
        if (options.baseUrl() != null) {
            return validateBaseUrl(options.baseUrl());
        }

        return environmentOrDefault(options.environment()).baseUrl();
    }

    private static String validateBaseUrl(String baseUrl) {
        if (baseUrl.trim().isEmpty()) {
            throw new PactmanConfigurationException("`baseUrl` must be a non-empty URL string.");
        }

        URI parsed;

        try {
            parsed = new URI(baseUrl.trim());
        } catch (URISyntaxException malformed) {
            throw new PactmanConfigurationException(
                    "`baseUrl` is not a valid URL: \"" + baseUrl
                            + "\". Expected something like https://entities.pactman.org.");
        }

        String scheme = parsed.getScheme() == null
                ? null
                : parsed.getScheme().toLowerCase(java.util.Locale.ROOT);

        if (!"https".equals(scheme) && !"http".equals(scheme)) {
            throw new PactmanConfigurationException(
                    "`baseUrl` must use http or https, received \"" + parsed.getScheme() + "\".");
        }

        if (parsed.getHost() == null) {
            throw new PactmanConfigurationException(
                    "`baseUrl` is missing a host: \"" + baseUrl + "\".");
        }

        String path = parsed.getRawPath() == null ? "" : parsed.getRawPath();

        while (path.endsWith("/")) {
            path = path.substring(0, path.length() - 1);
        }

        return scheme + "://" + parsed.getRawAuthority() + path;
    }

    private static long resolveTimeout(Long timeoutMs) {
        if (timeoutMs == null) {
            return PactmanClientOptions.DEFAULT_TIMEOUT_MS;
        }

        if (timeoutMs <= 0) {
            throw new PactmanConfigurationException(
                    "`timeoutMs` must be greater than zero. There is no way to disable the "
                            + "timeout.");
        }

        return timeoutMs;
    }

    private static Double resolveRequestsPerSecond(Double value) {
        if (value == null) {
            return null;
        }

        if (value.isNaN() || value.isInfinite() || value <= 0) {
            throw new PactmanConfigurationException(
                    "`maxRequestsPerSecond` must be a finite number greater than zero, or "
                            + "omitted.");
        }

        return value;
    }

    private static HttpExchange resolveExchange(PactmanClientOptions options) {
        if (options.httpExchange() != null) {
            return options.httpExchange();
        }

        if (options.httpClient() != null) {
            return new JdkHttpExchange(options.httpClient());
        }

        return JdkHttpExchange.withDefaults();
    }

    /** The API key and the resolved configuration, held apart on purpose. */
    public static final class Resolved {
        private final String apiKey;
        private final ClientConfig config;

        private Resolved(String apiKey, ClientConfig config) {
            this.apiKey = apiKey;
            this.config = config;
        }

        /**
         * The validated API key.
         *
         * @return the key, trimmed.
         */
        public String apiKey() {
            return apiKey;
        }

        /**
         * The resolved configuration, which does not carry the key.
         *
         * @return the configuration.
         */
        public ClientConfig config() {
            return config;
        }
    }
}
